Improve TLS codegen by marking the panic/init path as cold

orlp opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/143511)*
<!-- homu-ignore:end -->

This is an extension of the performance improvements seen from <https://github.com/rust-lang/rust/pull/141685>. I noticed that the non-`const` TLS still didn't have the `#[cold]` attribute for the uninit/panic path, and I also realized that neither implementation should have the initialization or panic path inlined, ever.

These paths are taken either only once per thread (`init`) or never (`panic`, in a well-behaving Rust program), thus they don't deserve to litter the code generated each time you access a thread-local variable. So in addition to `#[cold]` I added the more aggressive `#[inline(never)]` to both cold paths as well.

orlp committed:
Improve TLS codegen by marking the panic/init path as cold

compiler-errors commented:
Not sure if this will show up at all on perf but 🤷 

@bors2 try @rust-timer queue

Do you have any local benchmarks?

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

orlp committed:
Also apply opt to OS-specific TLS impls

orlp commented:
@compiler-errors No I don't have any local benchmarks. But I look at assembly output a lot, and trust me when I say these code paths should *never* get inlined.

Could you restart the benchmark with my second commit included?

compiler-errors commented:
@bors2 try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

orlp committed:
Don't use inline(never)

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (9782d0a1d99759de86b20e0863061637a0a3c245): [comparison URL](https://perf.rust-lang.org/compare.html?start=c83e217d268d25960a0c79c6941bcb3917a6a0af&end=9782d0a1d99759de86b20e0863061637a0a3c245&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=c83e217d268d25960a0c79c6941bcb3917a6a0af&end=9782d0a1d99759de86b20e0863061637a0a3c245&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c83e217d268d25960a0c79c6941bcb3917a6a0af&end=9782d0a1d99759de86b20e0863061637a0a3c245&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=c83e217d268d25960a0c79c6941bcb3917a6a0af&end=9782d0a1d99759de86b20e0863061637a0a3c245&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=c83e217d268d25960a0c79c6941bcb3917a6a0af&end=9782d0a1d99759de86b20e0863061637a0a3c245&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 9     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | 0.0%  | [0.0%, 0.0%]   | 1     |
</details>

**Bootstrap**: 461.809s -> 462.209s (0.09%)
**Artifact size**: 372.19 MiB -> 372.13 MiB (-0.02%)

ibraheemdev commented on library/std/src/sys/thread_local/native/eager.rs:
```diff
@@ -30,16 +30,22 @@ impl<T> Storage<T> {
     }
 
     #[cold]
-    unsafe fn initialize(&self) -> *const T {
```
Should this be marked as `#[inline(never)]`? I've noticed `destructors::register` calls being inlined into the hot-path before.

joboet commented on library/std/src/sys/thread_local/native/eager.rs:
```diff
@@ -4,7 +4,7 @@ use crate::sys::thread_local::{abort_on_dtor_unwind, destructors};
 #[derive(Clone, Copy)]
 enum State {
-    Initial,
+    Uninitialized,
```
`Uninitialized` isn't a great name, the value itself is very much initialised. It's just the destructor that's not registered yet.

orlp replied:
I've renamed this to `Unregistered`.

joboet commented on library/std/src/sys/thread_local/os.rs:
```diff
@@ -68,6 +68,7 @@ impl<T: 'static> Storage<T> {
     /// The resulting pointer may not be used after reentrant inialialization
     /// or thread destruction has occurred.
+    #[inline]
     pub fn get(&'static self, i: Option<&mut Option<T>>, f: impl FnOnce() -> T) -> *const T {
```
While you're at it, I think it might be beneficial to inline the `ptr.addr() == 1` case into this function, as that might yield more optimized `LocalKey::with`s.

orlp replied:
I disagree, see [my other comment](https://github.com/rust-lang/rust/pull/143511#discussion_r3362151647).

joboet commented on library/std/src/sys/thread_local/native/eager.rs:
```diff
@@ -30,16 +30,22 @@ impl<T> Storage<T> {
-            State::Alive => self.val.get(),
-            State::Destroyed => ptr::null(),
-            State::Initial => unsafe { self.initialize() },
+        if let State::Alive = self.state.get() {
+            self.val.get()
+        } else {
+            unsafe { self.get_or_init_slow() }
```
I don't think this is beneficial – the returned pointer is later compared against null in `LocalKey::with` anyway, so the optimiser should be able to merge the state comparison into that.

orlp replied:
I think it is beneficial. I think anything that is not the initialized path should be marked cold and gotten out of the way, even if that makes the non-initialized path slightly slower and have duplicated work.

The initialized hot path is what matters 99.999% of the time and should be prioritized over all else.

orlp replied:
I've made this toy example to illustrate this: <https://rust.godbolt.org/z/hd6hnGWGh>.

Note that because `UnsafeCell::get` cannot return a null pointer, the fast-path once inlined completely eliminates the nullptr check and only checks the state.

workingjubilee commented:
whoops, didn't mean for this to slip, but I espy a nominee

r? @joboet

orlp committed:
Rename Uninitialized to Unregistered

orlp committed:
Use zero to represent alive state for faster comparison on Arm

orlp commented:
Sorry for the delay. I've rebased on the latest main and addressed the review comments (either solving their concern or contesting).

I've also made one small change, I've explicitly assigned discriminant 0 to the `Alive` state which can give faster code on Arm to check (directly `cbz` instead of `cmp` first).

@rustbot label -S-waiting-on-author +S-waiting-on-review

joboet commented:
I tested the performance of my idea, and it truly appears to be worse.

Let's do a final perf-run of this, and then this should be good to go.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued 4fa024445d244dbe0860ff6c1500d718a02ec239 with parent 43a4909ee98ed4d006d9d773f5d94dc58e34f846, future [comparison URL](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=4fa024445d244dbe0860ff6c1500d718a02ec239).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (4fa024445d244dbe0860ff6c1500d718a02ec239): [comparison URL](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=4fa024445d244dbe0860ff6c1500d718a02ec239&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking means the PR may be perf-sensitive. Consider adding rollup=never if this change is not fit for rolling up.


@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=4fa024445d244dbe0860ff6c1500d718a02ec239&stat=instructions:u)
This perf run didn't have relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=4fa024445d244dbe0860ff6c1500d718a02ec239&stat=max-rss)
<details>
<summary>Results (secondary -4.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.9% | [-4.9%, -4.9%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=4fa024445d244dbe0860ff6c1500d718a02ec239&stat=cycles:u)
This perf run didn't have relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=43a4909ee98ed4d006d9d773f5d94dc58e34f846&end=4fa024445d244dbe0860ff6c1500d718a02ec239&stat=size:linked_artifact)
This perf run didn't have relevant results for this metric.

**Bootstrap**: 516.093s -> 514.238s (-0.36%)
**Artifact size**: 400.83 MiB -> 401.34 MiB (0.13%)

joboet commented:
Well, that's underwhelming...

But I think I'll merge this anyway, it makes the implementations more consistent and is nice to have in general...

@bors r+ rollup

JonathanBrouwer commented:
@rust-timer build 6f652bccf0064dcdc88d6c205623d026ea512f8d

rust-timer commented:
Queued 6f652bccf0064dcdc88d6c205623d026ea512f8d with parent f3ef3bd882dd24a275a60701a67c3bb330edd8c1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f3ef3bd882dd24a275a60701a67c3bb330edd8c1&end=6f652bccf0064dcdc88d6c205623d026ea512f8d).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (6f652bccf0064dcdc88d6c205623d026ea512f8d): [comparison URL](https://perf.rust-lang.org/compare.html?start=f3ef3bd882dd24a275a60701a67c3bb330edd8c1&end=6f652bccf0064dcdc88d6c205623d026ea512f8d&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking means the PR may be perf-sensitive. Consider adding rollup=never if this change is not fit for rolling up.


@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=f3ef3bd882dd24a275a60701a67c3bb330edd8c1&end=6f652bccf0064dcdc88d6c205623d026ea512f8d&stat=instructions:u)
This perf run didn't have relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f3ef3bd882dd24a275a60701a67c3bb330edd8c1&end=6f652bccf0064dcdc88d6c205623d026ea512f8d&stat=max-rss)
<details>
<summary>Results (primary -1.3%, secondary 6.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.8%, 2.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 6.8%  | [6.8%, 6.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -5.4% | [-5.4%, -5.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.3% | [-5.4%, 2.8%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f3ef3bd882dd24a275a60701a67c3bb330edd8c1&end=6f652bccf0064dcdc88d6c205623d026ea512f8d&stat=cycles:u)
<details>
<summary>Results (secondary -0.5%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.6%  | [3.6%, 3.6%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.6% | [-4.6%, -4.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f3ef3bd882dd24a275a60701a67c3bb330edd8c1&end=6f652bccf0064dcdc88d6c205623d026ea512f8d&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0% | [0.0%, 0.0%] | 3     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.0% | [0.0%, 0.0%] | 3     |
</details>

**Bootstrap**: 517.572s -> 515.951s (-0.31%)
**Artifact size**: 400.85 MiB -> 400.77 MiB (-0.02%)
